Yet another sharding support implementation

clslrns opened the pull request:
This one is inspired by #691.

Brief summary:
* `shardKey` property was added to `ClassMetadataInfo`, mapping support were added to all drivers. It's a hash with `keys` and `options`. `keys` was originally called `fields`, but I decided to conform to the indexes definition.
* `shardKey` is class-based, since we need to keep the order of the keys to ensure compound index.
* `ensureSharding` method were added to `SchemaManager`. It's similar to `ensureIndex`. For non-empty collections it will run `ensureIndex` and retry `shardCollection` again.
* `DocumentPersister` uses `shardKey` to generate query part for `upsert`, `update`, `remove`, `findOne`.

I tried to cover all the new code with both unit and functional tests. Tests were run against sharded clusters of MongoDB 2.2.3, 2.4.9 and 2.6.5. Functional tests were tagged to avoid running against non-sharded setup.

Now I'm testing it on production. I'd like to receive some feedback from you guys to improve or change something.

clslrns commented:
It seems, tests failed because of long MongoDB initialization. Would be great if somebody re-ran them.

jmikola commented:
@clslrns: I will attempt to get to this before the next beta tag.

clslrns commented:
Btw, wouldn't it be better to squash all this commits?

notrix commented:
:+1:

alcaeus commented:
@clslrns sorry for the long time without updates. Could you please rebase the branch to see if it still works against the current master?

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -422,13 +429,12 @@ public function delete($document, array $options = array())
      * @param object $document The document to refresh.
      */
-    public function refresh($id, $document)
+    public function refresh($document)
```
I'm afraid this is a BC break. As useless as the ID parameter has become, it has to stay. Feel free to deprecate it.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -348,14 +352,19 @@ private function executeUpsert(array $data, array $options)
-            $id = $this->class->getDatabaseIdentifierValue($id);
-            $query = array('_id' => $id);
+        foreach (array_keys($query) as $field) {
+            unset($update['$set'][$field]);
```
From the looks of it, this could pose issues where a field with an ```@increment``` mapping is part of the shard key. The field would not be part of a ```$set``` operator but an ```$inc``` operator, meaning that we'd send an invalid query to the database.
In this case, however, I think we should set limits on what kind of mapping can be used for fields that are part of the shard key. Having an ```@increment``` mapping makes no sense since the field should not be updated anyways, so I suggest guarding against this on the ClassMetadata level.

clslrns commented:
@alcaeus, thanks for the feedback.
I have rebased it and fixed two issues you pointed to. I hope, in a few days I'll test rebased version against the app.

darceee2 commented:
:+1:

alcaeus commented:
Superseded by #1385.
